book/src/rfcs/ffadd.md

```diff
@@ -0,0 +1,194 @@
+with a field overflow term $q$ that will be either $0$ or $1$. Looking at this in limb form, we have:
+
+```text
+bits  0..............87|88...........175|176...........263
+
+a  =  (-------a0-------|-------a1-------|-------a2-------)
++
+b  =  (-------b0-------|-------b1-------|-------b2-------)
+=
+q  =  0 | 1
+·
+f  =  (-------f0-------|-------f1-------|-------f2-------)
++
+r  =  (-------r0-------|-------r1-------|-------r2-------)
+```
+
+First, if $a + b$ was larger than $f$, then we will have a field overflow (represented by $q = 1$) and we will have to subtract $f$ from the sum $s = a + b$ to obtain $r$.  Whereas the foreign field overflow necessitates a overflow bit $q$ for the foreign field equation above, when $q = 1$ there is a corresponding subtraction that may introduce carries (or even borrows) between the limbs.  This is because $r = a + b - q \cdot f \mod 2^{264}$.  Therefore, in the equations for the limbs we will use a carry flag $c_i$ for limb $i$ to represent both carries and borrows.  The carry flags can be anything in $\{-1, 0, 1\}$, where $c_i = -1$ represents a borrow.  Next we see more clearly how this works.
+
+In order to perform this operation in parts, we first take a look at the least significant limb, which is the easiest part. This means we want to know how to compute $r_0$. First, if the addition of the bits in $a_0$ and $b_0$ produce a carry bit, then it should propagate to the second limb. That means one has to subtract $2^{88}$ from $s_0 = a_0 + b_0$, add 1 to $s_1 = a_1 + b_1$ and set the low carry flag $c_0$ to 1 (otherwise it is zero).  Thus, the equation for the lowest bit is
+
+$$a_0 + b_0 = q \cdot f_0 + r_0 + c_0 \cdot 2^{88}$$
+
+Or put in another way, this is equivalent to saying that $a_0 + b_0 - q \cdot f_0 - r_0$ is a multiple of $2^{88}$ (or, the existence of the carry coefficient $c_0$).
+
+This kind of equation needs an additional check that the carry coefficient $c_0$ is a binary value. We will use this idea for the remaining limbs as well.
+
+Looking at the second limb, we first need to observe that the addition of $a_1$ and $b_1$ can, not only produce a carry bit $c_1$, but they may need to take into account the carry bit from the first limb; $c_0$. Similarly to the above,
+
+$$a_1 + b_1 = q \cdot f_1 + r_1 + c_1 \cdot 2^{88} - c_0$$
+
+Note that in this case, the carry coefficient from the least significant limb is not multiplied by $2^{88}$, but instead is used as is, since it occupies the least significant position of the second limb. Here, the second carry bit $c_1$ is the one being followed by $88$ zeros. Again, we will have to check that $c_1$ is a bit.
+
+Finally, for the third limb, we obtain a similar equation. But in this case, we do not have to take into account $c_0$ anymore, since it was already considered within $c_1$. Again, the most significant carry bit $c_2$ should be a bit.
+
+$$a_2 + b_2 = q \cdot f_2 + r_2 + c_2 \cdot 2^{88} - c_1$$
```

Comment:
Maybe we need a note about not needing `c_2 c_2 \cdot 2^{88}` since its a multiple of `2^{264}`

Reply:
yeah this is old version, c_2 should always be zero as we pointed out above